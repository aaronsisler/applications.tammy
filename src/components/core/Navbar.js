import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { startLogout } from '../../actions/auth';

const Navbar = (props) => (
    <div id="navbar">
        {props.isAuthenticated &&
            <div className="navbar_content">
                <div className="navbar_links">
                    <Link className="navbar_link" to="/dashboard">Dashboard</Link>
                    <Link className="navbar_link" to="/user_account">User Account</Link>
                    <Link className="navbar_link" to="/">Positions</Link>
                </div>
                <div className="navbar_logout">
                    <button className="button navbar_link" onClick={props.startLogout}>Logout</button>
                </div>
            </div>
        }
        {!props.isAuthenticated &&
            <div className="navbar_content">
                <div className="navbar_login">
                    <Link className="navbar_link" to="/login">Login</Link>
                </div>
            </div>
        }
    </div>
);

Navbar.propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    startLogout: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
    startLogout: () => dispatch(startLogout())
})

export default connect(undefined, mapDispatchToProps)(Navbar);
