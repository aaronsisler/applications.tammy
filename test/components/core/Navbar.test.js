import React from 'react';
import { shallow } from 'enzyme';
import { Navbar } from 'Core/Navbar';
import {
    bodyAddMock,
    bodyRemoveMock,
    getElementByIdClassListMock,
} from '../../__mocks__/document';

describe('Navbar', () => {
    const startClearUser = jest.fn();
    const startLogout = jest.fn();
    const bodyOpenClassName = 'body_open';
    const navSideMenuOpenClassName = 'nav_side_menu_open';
    const navSideMenu = 'nav_side_menu';
    let wrapper;
    let instance;

    beforeEach(() => {
        // console.log('beforeEach main');
        wrapper = shallow(
            <Navbar
                isAuthenticated={true}
                startClearUser={startClearUser}
                startLogout={startLogout}
            />);
        instance = wrapper.instance();
    });

    afterEach(() => {
        // console.log('Restore mocks');
        jest.restoreAllMocks();
    });

    it('should have isNavOpen set to false on initialization', () => {
        expect(wrapper.state('isNavOpen')).toBe(false);
    });

    describe('favicons', () => {
        beforeEach(() => {
            // console.log('beforeEach spyOn');
            jest.spyOn(instance, 'handleOpenSideMenu').mockImplementation(jest.fn());
            jest.spyOn(instance, 'handleCloseSideMenu').mockImplementation(jest.fn());
        });

        describe('times icon', () => {
            it('should show if isNavOpen is TRUE', () => {
                wrapper.setState({ isNavOpen: true });
                expect(wrapper.state('isNavOpen')).toBe(true);
                expect(wrapper.find('.nav_toggles').children().prop('icon').iconName).toBe('times');
            });

            it('should call handleCloseSideMenu when clicked', () => {
                wrapper.setState({ isNavOpen: true });
                expect(wrapper.state('isNavOpen')).toBe(true);
                wrapper.find('.navbar_favicon').simulate('click');
                expect(instance.handleCloseSideMenu).toHaveBeenCalled();
            });
        });

        describe('bars icon', () => {
            it('should show if isNavOpen is FALSE', () => {
                expect(wrapper.state('isNavOpen')).toBe(false);
                expect(wrapper.find('.nav_toggles').children().prop('icon').iconName).toBe('bars');
            });

            xit('should call handleOpenSideMenu when clicked', () => {
                // console.log('In test');
                expect(wrapper.state('isNavOpen')).toBe(false);
                expect(wrapper.find('.navbar_favicon')).toHaveLength(1);
                // console.log('After first 2 expects');
                wrapper.find('.navbar_favicon').simulate('click');
                expect(instance.handleOpenSideMenu).toHaveBeenCalled();
                // console.log('End of test');
            });
        });
    });

    describe('handleOpenSideMenu() method', () => {
        beforeEach(() => {
            jest.spyOn(document.body.classList, 'add').mockImplementation(bodyAddMock);
            jest.spyOn(document, 'getElementById').mockReturnValue(getElementByIdClassListMock);
        });

        it('should add classname to body classlist', () => {
            instance.handleOpenSideMenu();
            expect(bodyAddMock).toHaveBeenLastCalledWith(bodyOpenClassName);
            expect(document.body.classList.add).toHaveBeenLastCalledWith(bodyOpenClassName);
        });

        it(`should add classname to nav side menu element's classlist`, () => {
            instance.handleOpenSideMenu();
            expect(document.getElementById).toHaveBeenLastCalledWith(navSideMenu);
            expect(document.getElementById(navSideMenu).classList.add).toHaveBeenLastCalledWith(navSideMenuOpenClassName);
        });

        it('should set isNavOpen in state to true', () => {
            instance.handleOpenSideMenu();
            expect(wrapper.state('isNavOpen')).toBe(true);
        });
    })

    describe('handleCloseSideMenu() method', () => {
        beforeEach(() => {
            jest.spyOn(document.body.classList, 'remove').mockImplementation(bodyRemoveMock);
            jest.spyOn(document, 'getElementById').mockReturnValue(getElementByIdClassListMock);
        });

        it('should remove classname to body classlist', () => {
            instance.handleCloseSideMenu();
            expect(bodyRemoveMock).toHaveBeenLastCalledWith(bodyOpenClassName);
            expect(document.body.classList.remove).toHaveBeenLastCalledWith(bodyOpenClassName);
        });

        it(`should remove classname to nav side menu element's classlist`, () => {
            instance.handleCloseSideMenu();
            expect(document.getElementById).toHaveBeenLastCalledWith(navSideMenu);
            expect(document.getElementById(navSideMenu).classList.remove)
                .toHaveBeenLastCalledWith(navSideMenuOpenClassName);
        });

        it('should set isNavOpen in state to false', () => {
            expect(wrapper.state('isNavOpen')).toBe(false);
            wrapper.setState({ isNavOpen: true });
            expect(wrapper.state('isNavOpen')).toBe(true);
            instance.handleCloseSideMenu();
            expect(wrapper.state('isNavOpen')).toBe(false);
        });
    })

    describe('when user is authenticated', () => {
        it('should render Navbar correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });

        describe('handleLogout() method', () => {
            beforeEach(() => {
                jest.spyOn(instance, 'handleOpenSideMenu').mockImplementation(jest.fn());
                jest.spyOn(instance, 'handleCloseSideMenu').mockImplementation(jest.fn());
            })

            it('should call handleCloseSideMenu', () => {
                wrapper.find('button').simulate('click');

                expect(instance.handleCloseSideMenu).toHaveBeenCalled();
            })

            it('should call startClearUser', () => {
                wrapper.find('button').simulate('click');

                expect(startClearUser).toHaveBeenCalled();
            })

            it('should call startLogout', () => {
                wrapper.find('button').simulate('click');

                expect(startLogout).toHaveBeenCalled();
            })
        })
    })

    describe('when user is NOT authenticated', () => {
        beforeEach(() => {
            wrapper = shallow(
                <Navbar
                    isAuthenticated={false}
                    startClearUser={startClearUser}
                    startLogout={startLogout}
                />)
        });

        it('should render Navbar correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    })
})
