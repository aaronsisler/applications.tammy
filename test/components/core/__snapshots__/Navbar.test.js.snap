// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar when user is NOT authenticated should render Navbar correctly 1`] = `
<div
  id="navbar"
>
  <div
    className="nav_toggles"
  >
    <FontAwesomeIcon$1
      border={false}
      className="navbar_favicon"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f0c9",
            "M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z",
          ],
          "iconName": "bars",
          "prefix": "fas",
        }
      }
      listItem={false}
      mask={null}
      name=""
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size="2x"
      spin={false}
      symbol={false}
      transform={null}
    />
  </div>
  <div
    id="nav_side_menu"
  >
    <Link
      className="nav_side_menu_link"
      onClick={[Function]}
      replace={false}
      to="/"
    >
      Positions
    </Link>
    <Link
      className="nav_side_menu_link"
      onClick={[Function]}
      replace={false}
      to="/login"
    >
      Login
    </Link>
  </div>
</div>
`;

exports[`Navbar when user is authenticated should render Navbar correctly 1`] = `
<div
  id="navbar"
>
  <div
    className="nav_toggles"
  >
    <Connect(NotificationsContainer) />
    <FontAwesomeIcon$1
      border={false}
      className="navbar_favicon"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f0c9",
            "M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z",
          ],
          "iconName": "bars",
          "prefix": "fas",
        }
      }
      listItem={false}
      mask={null}
      name=""
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size="2x"
      spin={false}
      symbol={false}
      transform={null}
    />
  </div>
  <div
    id="nav_side_menu"
  >
    <Link
      className="nav_side_menu_link"
      onClick={[Function]}
      replace={false}
      to="/dashboard"
    >
      Dashboard
    </Link>
    <Link
      className="nav_side_menu_link"
      onClick={[Function]}
      replace={false}
      to="/user_profile"
    >
      User Profile
    </Link>
    <Link
      className="nav_side_menu_link"
      onClick={[Function]}
      replace={false}
      to="/user_documents"
    >
      User Documents
    </Link>
    <Link
      className="nav_side_menu_link"
      onClick={[Function]}
      replace={false}
      to="/"
    >
      Positions
    </Link>
    <button
      className="button nav_side_menu_link"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
</div>
`;
